nixos/modules/services/matrix/synapse/workers.nix

```diff
@@ -0,0 +1,198 @@
+                  "static"
+                  "webclient" # TODO: wtf?
+                ]);
+                description = "A list of resources to host on this port";
+                default = [ ];
+                example = lib.literalExpression ''[ "client" ]'';
+              };
+              compress = lib.mkEnableOption "HTTP compression for this resource";
+            };
+          });
+          default = [ { } ];
+          description = "The resources to make available on this listener";
+        };
+      };
+    };
+
+  in {
+    mainReplicationHost = mkOption {
+      type = types.str;
+      default = let
+        host = lib.head mainReplicationListener.bind_addresses;
+      in if builtins.elem host [ "0.0.0.0" "::"] then "127.0.0.1" else host;
+      defaultText = "Autoconfigured from main synapse config";
+      description = "The replication listeners IP on the main synapse process";
+    };
+
+    mainReplicationPort = mkOption {
+      type = types.port;
+      default = mainReplicationListener.port;
+      defaultText = "Autoconfigured from main synapse config";
+      description = "The replication listeners port on the main synapse process";
+    };
+
+    instances = mkOption {
+      type = types.attrsOf workerInstanceType;
+      default = { };
+      example = {
+        "federation_sender1" = {
+          settings = {
+            worker_name = "federation_sender1";
+            worker_app = "synapse.app.generic_worker";
+
+            worker_replication_host = "127.0.0.1";
+            worker_replication_http_port = 9093;
+            worker_listener = [ ];
+          };
+        };
+      };
+      description = "The name of the worker and its configuration";
+    };
+  };
+
+  config = {
+    systemd.services = let
+      workerList = lib.mapAttrsToList lib.nameValuePair cfg.workers.instances;
+      workerConfig = worker: format.generate "matrix-synapse-worker-${worker.name}-config.yaml"
+        (worker.value.settings // { worker_name = worker.name; });
+      in builtins.listToAttrs (lib.flip map workerList (worker: {
+        name = "matrix-synapse-worker-${worker.name}";
+        value = {
+          description = "Matrix Synapse Worker";
+          partOf = [ "matrix-synapse.target" ];
+          wantedBy = [ "matrix-synapse.target" ];
+          after = [ "matrix-synapse.service" ];
+          requires = [ "matrix-synapse.service" ];
+          environment.PYTHONPATH = lib.makeSearchPathOutput "lib" cfg.package.python.sitePackages [
+            pluginsEnv
+          ];
+          serviceConfig = {
+            Type = "notify";
+            User = "matrix-synapse";
+            Group = "matrix-synapse";
+            Slice = "system-matrix-synapse.slice";
+            WorkingDirectory = cfg.dataDir;
+            ExecStartPre = pkgs.writers.writeBash "wait-for-synapse" ''
+              while ! systemctl is-active -q matrix-synapse.service; do
+                  sleep 1
+              done
+            '';
```

Comment:
This shouldn't be necessary when using `after = [ "matrix-synapse.service" ]`.

Reply:
This is for when nixos-switch runs `systemctl restart`. systemd for some reason does not respect after when _restarting_ a service.

Reply:
Having sleep loops in startup scripts is rather ugly and should be avoided at all cost. You are absolutely sure that systemd has nothing to solve this? Also isn't nixos-rebuild by default stopping and starting services?

Reply:
We didn't use to have this in the module this was ported from, and there were loads of problems.

The only systemd native way I could imagine fixing this would be setting the restart options to keep retrying after failure indefinetly. But that's IMO an even worse solution than this